### PR TITLE
Authorisation

### PR DESCRIPTION
**Pull request by siapankina**
Story: https://trello.com/c/G0siEEpZ/23-as-a-contract-manager-i-can-approve-or-reject-a-variation-so-that-contractors-know-what-to-do-next

Screenshot 2021-04-12 at 17 49 20

### CODE REVIEW

File: src/pages/repairs/jobs/[id]/authorisation.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  return {
+    props: {
+      query,
+    },
+  }
+}
+
+AuthorisationPage.permittedRoles = [
+  AGENT_ROLE,
+  CONTRACTOR_ROLE,
```
[the remark this answers is not in the document]
Comment (benbaumann95) [reply src/pages/repairs/jobs/[id]/authorisation.js:23]: Does an agent have access to this page?

[the remark this answers is not in the document]
Comment (EGiataganas) [reply src/pages/repairs/jobs/[id]/authorisation.js:23]: Also can you please add a unit test for this? @siapankina

File: src/styles/components/work-orders.scss [hunk 2]
```diff
@@ -49,5 +49,8 @@
     .status-pending-approval {
       background-color: repairs-hub-colour('status-waiting');
     }
+    .status-variation-approved {
+      background-color: repairs-hub-colour('dark-green');
+    }
   }
```
Comment (EGiataganas) [inline src/styles/components/work-orders.scss:55]: Do you need to add a colour when the status variation is rejected?

File: src/components/WorkOrder/Authorisation/AuthorisationView.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      {loading ? (
+        <Spinner />
+      ) : (
+        <div>
+          <BackButton />
+          <h1 className="govuk-heading-l">
+            Authorisation request: {workOrderReference}{' '}
+          </h1>
+          <Link href={`/work-orders/${workOrderReference}`}>
+            <a>See works order</a>
```
Comment (benbaumann95) [inline src/components/WorkOrder/Authorisation/AuthorisationView.js:66]: From the screenshot I think we need to apply some govuk styling to this

File: src/utils/hact/job-status-update/authorisation.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+export const buildAuthorisationApprovedFormData = (
```
[the remark this answers is not in the document]
Comment (siapankina) [reply src/utils/hact/job-status-update/authorisation.js:1]: 1- hey Ben , @Evangelos suggested to have 2 separate functions, for readability 
2- yes, backend autopopulate note when variation is approved

[the remark this answers is not in the document]
Comment (EGiataganas) [reply src/utils/hact/job-status-update/authorisation.js:1]: Hey @benbaumann95, here's the [code][1] where the backend adds an automated note when the variation is approved

[1]: https://github.com/LBHackney-IT/repairs-api-dotnet/blob/develop/RepairsApi/V2/UseCase/JobStatusUpdatesUseCases/ApproveVariationUseCase.cs#L33

[the remark this answers is not in the document]
Comment (benbaumann95) [reply src/utils/hact/job-status-update/authorisation.js:1]: Ah brilliant - looks great, thanks

File: cypress/integration/jobs/job-authorisation.spec.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      response: '',
+    }).as('apiCheck')
+  })
+
+  it('Rejects job variation', () => {
+    cy.visit(`${Cypress.env('HOST')}/work-orders/10000012`)
+    cy.contains('a', 'Update Works Order').click()
+    cy.contains('Authorisation request: 10000012')
+
+    // Throughs an error when rejected without comments
```
Comment (benbaumann95) [inline cypress/integration/jobs/job-authorisation.spec.js:26]: Hey just a little typo. Should be `throws an error ...`

File: cypress/integration/jobs/job-authorisation.spec.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        typeCode: '125',
+      })
+  })
+
+  it('Approves job variation', () => {
+    cy.visit(`${Cypress.env('HOST')}/work-orders/10000012`)
+    cy.contains('a', 'Update Works Order').click()
+    cy.contains('Authorisation request: 10000012')
+
+    // When click approve default commets appear
```
Comment (benbaumann95) [inline cypress/integration/jobs/job-authorisation.spec.js:54]: typo `comments`

File: cypress/support/commands.js [hunk 7]
```diff
@@ -46,6 +46,15 @@ Cypress.Commands.add('loginWithContractorRole', () => {
   cy.visit(host)
 })
 
+Cypress.Commands.add('loginWithContractManagerRole', () => {
```
Comment (benbaumann95) [inline cypress/support/commands.js:49]: 👍

File: src/components/WorkOrder/Authorisation/AuthorisationView.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
+import PropTypes from 'prop-types'
+import Spinner from '../../Spinner/Spinner'
+import ErrorMessage from '../../Errors/ErrorMessage/ErrorMessage'
+import { useState } from 'react'
+import BackButton from '../../Layout/BackButton/BackButton'
+import Radios from '../../Form/Radios/Radios'
+import TextArea from '../../Form/TextArea/TextArea'
+import { PrimarySubmitButton } from '../../Form'
```
Comment (benbaumann95) [inline src/components/WorkOrder/Authorisation/AuthorisationView.js:8]: You could import all this in one statement e.g. 
```
import {
  Radios,
  TextArea,
  PrimarySubmitButton
} from '../../Form'

File: src/utils/hact/job-status-update/authorisation.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  formData,
+  workOrderReference
+) => {
+  return {
+    relatedWorkOrderReference: {
+      id: workOrderReference,
+    },
+    typeCode: '100-20',
```
[the remark this answers is not in the document]
Comment (benbaumann95) [reply src/utils/hact/job-status-update/authorisation.js:7]: Can we have the actual jobStatusUpdateTypeCode description from the excel HACT doc i.e. `Additional work approved by supervisor`

File: src/utils/hact/job-status-update/authorisation.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+export const buildAuthorisationRejectedFormData = (
+  formData,
+  workOrderReference
+) => {
+  return {
+    relatedWorkOrderReference: {
+      id: workOrderReference,
+    },
+    comments: formData.note,
+    typeCode: '125',
```
[the remark this answers is not in the document]
Comment (benbaumann95) [reply src/utils/hact/job-status-update/authorisation.js:21]: This is a custom code right ? 
Can we indicate this in our comment  and just a tiny nitpick here we need a space between the start of comment `//` and `code`

**Review by benbaumann95 — APPROVED** [review]
Just a few minor comments to address please. 
Otherwise looks great, well done! 🎉

File: cypress/fixtures/repairs/work-orders-pending-approval.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    "priority": "U - Urgent 7 days (5 Working days)",
+    "property": "315 Banister House  Homerton High Street",
+    "owner": "Alphatrack (S) Systems Lt",
+    "description": "A very urgent repair",
+    "propertyReference": "00012345",
+    "tradeCode": "DE",
+    "tradeDescription": "DOOR ENTRY ENGINEER - DE",
+    "status": "Pending approval"
+  }
+]
```
Comment (EGiataganas) [inline cypress/fixtures/repairs/work-orders-pending-approval.json:28]: where do you use this file?

**Review by EGiataganas — APPROVED** [review]
I've just have one question, other than that it looks great! Well done :shipit:

**Pull request merged**